src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1857,22 +1857,28 @@ private IEnumerable<TypeInferenceInfo> InferTypeForReturnStatement(ReturnStateme
                     return;
                 }
 
-                var ancestorExpressions = returnStatement.GetAncestorsOrThis<ExpressionSyntax>();
+                var ancestor = returnStatement.AncestorsAndSelf().FirstOrDefault(e => e is ParenthesizedLambdaExpressionSyntax || e is SimpleLambdaExpressionSyntax ||
+                    e is AnonymousMethodExpressionSyntax || e is LocalFunctionStatementSyntax);
 
                 // If we're in a lambda, then use the return type of the lambda to figure out what to
                 // infer.  i.e.   Func<int,string> f = i => { return Goo(); }
-                var lambda = ancestorExpressions.FirstOrDefault(e => e.IsKind(SyntaxKind.ParenthesizedLambdaExpression, SyntaxKind.SimpleLambdaExpression));
-                if (lambda != null)
+                if (ancestor is ParenthesizedLambdaExpressionSyntax lambda)
                 {
                     types = InferTypeInLambdaExpression(lambda);
-                    isAsync = lambda is ParenthesizedLambdaExpressionSyntax && ((ParenthesizedLambdaExpressionSyntax)lambda).AsyncKeyword.Kind() != SyntaxKind.None;
+                    isAsync = lambda.AsyncKeyword.Kind() != SyntaxKind.None;
+
                     return;
+
                 }
+                else if (ancestor is SimpleLambdaExpressionSyntax simpleLambda)
+                {
+                    types = InferTypeInLambdaExpression(simpleLambda);
 
-                // If we are inside a delegate then use the return type of the Invoke Method of the delegate type
-                var delegateExpression = (AnonymousMethodExpressionSyntax)ancestorExpressions.FirstOrDefault(e => e.IsKind(SyntaxKind.AnonymousMethodExpression));
-                if (delegateExpression != null)
+                    return;
+                }
+                else if (ancestor is AnonymousMethodExpressionSyntax delegateExpression)
                 {
+                    // If we are inside a delegate then use the return type of the Invoke Method of the delegate type
                     var delegateType = InferTypes(delegateExpression).FirstOrDefault().InferredType;
```

Comment:
This could likely be simplified too, because lambdas and anon. methods share all this info in `AnonymousFunctionExpressionSyntax`, but meh... this whole mess needs a rewrite :smile:

Reply:
I agree, I'll leave that to you ;).